spaghetto/src/Spago/Psa.purs

```diff
@@ -0,0 +1,197 @@
+import Data.Codec.Argonaut as CA
+import Data.DateTime.Instant (toDateTime)
+import Data.Foldable (foldr)
+import Data.Set as Set
+import Data.String as Str
+import Effect.Exception as Exception
+import Effect.Now (now)
+import Effect.Ref as Ref
+import Foreign.Object as FO
+import Node.Encoding as Encoding
+import Node.FS.Aff as FSA
+import Node.FS.Perms (permsAll)
+import Node.FS.Stats as Stats
+import Node.FS.Sync as FSSync
+import Node.Path (dirname)
+import Spago.Core.Config as Core
+import Spago.Psa.Output (buildOutput)
+import Spago.Psa.Printer.Default as DefaultPrinter
+import Spago.Psa.Printer.Json as JsonPrinter
+import Spago.Psa.Types (PsaOutputOptions, ErrorCode, psaErrorCodec, psaResultCodec)
+import Spago.Purs as Purs
+
+type PsaArgs =
+  { libraryDirs :: Array String
+  , jsonErrors :: Boolean
+  , color :: Boolean
+  }
+
+defaultParseOptions :: PsaOptions
+defaultParseOptions =
+  { showSource: Core.ShowSourceCode
+  , stashFile: Nothing
+  , censorWarnings: false
+  , censorLib: false
+  , censorSrc: false
+  , censorCodes: Set.empty
+  , filterCodes: Set.empty
+  , statVerbosity: Core.CompactStats
+  , strict: false
+  }
+
+type PsaOptions =
+  { showSource :: Core.ShowSourceCode
+  , stashFile :: Maybe String
+  , censorWarnings :: Boolean
+  , censorLib :: Boolean
+  , censorSrc :: Boolean
+  , censorCodes :: Set ErrorCode
+  , filterCodes :: Set ErrorCode
+  , statVerbosity :: Core.StatVerbosity
+  , strict :: Boolean
+  }
+
+toOutputOptions :: PsaArgs -> PsaOptions -> PsaOutputOptions
+toOutputOptions { libraryDirs, color } options =
+  { color
+  , censorWarnings: options.censorWarnings
+  , censorLib: options.censorLib
+  , censorSrc: options.censorSrc
+  , censorCodes: options.censorCodes
+  , filterCodes: options.filterCodes
+  , statVerbosity: options.statVerbosity
+  , libraryDirs
+  , strict: options.strict
+  }
+
+psaCompile :: forall a. Set.Set FilePath -> Array String -> PsaArgs -> PsaOptions -> Spago (Purs.PursEnv a) Unit
+psaCompile globs pursArgs psaArgs options@{ showSource, stashFile } = do
+  let outputOptions = toOutputOptions psaArgs options
+  stashData <- case stashFile of
+    Just f -> readStashFile f
+    Nothing -> emptyStash
+
+  result <- Purs.compile globs (Array.snoc pursArgs "--json-errors")
+  let
+    result' = case result of
+      Left err -> { output: err.stdout, exitCode: err.exitCode, err: Just err }
+      Right success -> { output: success.stdout, exitCode: Just success.exitCode, err: Nothing }
+  arrErrorsIsEmpty <- forWithIndex (Str.split (Str.Pattern "\n") result'.output) \idx err ->
+    case jsonParser err >>= CA.decode psaResultCodec >>> lmap CA.printJsonDecodeError of
+      Left decodeErrMsg -> do
+        logDebug $ Array.intercalate "\n"
+          [ "Failed to decode PsaResult at index '" <> show idx <> "': " <> decodeErrMsg
+          , "Json was: " <> err
+          ]
+        -- Note to reviewer:
+        -- Should a stash decode error cause a non-zero exit?
```

Comment:
The rule of thumb is "we crash if we can't figure out how to proceed". This seems an entirely recoverable situation, in the sense that we can in fact just move on without a stash file.
I would probably log a warning though (instead of a debug log), so we get bug reports - because a decoding failure is 95% likely to be our bug.